### ISSUE TITLE
RFC 9266: Channel Bindings for TLS 1.3 support

### ISSUE THREAD

**Issue by Neustradamus** [naqvis/cr-xmpp#9]
Dear @naqvis,

Can you add the support of RFC 9266: Channel Bindings for TLS 1.3?
- https://datatracker.ietf.org/doc/html/rfc9266

Channel Bindings for TLS: https://datatracker.ietf.org/doc/html/rfc5929

- XEP-0388: Extensible SASL Profile: https://xmpp.org/extensions/xep-0388.html
- XEP-0440: SASL Channel-Binding Type Capability: https://xmpp.org/extensions/xep-0440.html
- XEP-0474: SASL SCRAM Downgrade Protection: https://xmpp.org/extensions/xep-0474.html
- XEP-0480: SASL Upgrade Tasks: https://xmpp.org/extensions/xep-0480.html

Little details, to know easily:
- tls-unique for TLS =< 1.2 (RFC5929)
- tls-server-end-point =< 1.2 + 1.3 (RFC5929)
- tls-exporter for TLS = 1.3 (RFC9266)

After the jabber.ru MITM, it is time to add it:
- https://notes.valdikss.org.ru/jabber.ru-mitm/
- https://snikket.org/blog/on-the-jabber-ru-mitm/
- https://www.devever.net/~hl/xmpp-incident
- https://blog.jmp.chat/b/certwatch/certwatch

Thanks in advance.

Linked to:
- Channel Binding: https://github.com/scram-sasl/info/issues/1
- https://github.com/naqvis/cr-xmpp/issues/7

**Comment by Neustradamus** [comment]
@naqvis: Have you progressed on it?

**Comment by naqvis** [comment]
@Neustradamus Sorry don't have enough bandwidth for this change.

**Comment by Neustradamus** [comment]
@naqvis: Have you progressed on it?

**Comment by naqvis** [comment]
> [@naqvis](https://github.com/naqvis): Have you progressed on it?

nah, haven't touched the code base since long and don't have enough bandwidth to implement requested change.  Any PRs are welcome.

**Comment by naqvis** [comment]
@Neustradamus pushed release v0.2.0 that implemented requested changes, as well as clean-up to work with latest crystal version. You might find change log or release notes useful

**Comment by Neustradamus** [comment]
@naqvis: Very good job, thanks for your work!

It is excellent to see a complete support in cr-xmpp!

With SCRAM-SHA-1-PLUS, SCRAM-SHA-256-PLUS, SCRAM-SHA-512-PLUS too:
- https://github.com/naqvis/cr-xmpp/issues/7

In the README, you can change this part: "RFC 5929 - Channel Bindings for TLS (TLS 1.2 and earlier)".

Little details, to know easily:
- tls-unique for TLS =< 1.2 (RFC5929)
- tls-server-end-point =< 1.2 + 1.3 (RFC5929)
- tls-exporter for TLS = 1.3 (RFC9266)

Note: You have already a "good" part after.

Can you look for XEP-0480: SASL Upgrade Tasks support?
- https://xmpp.org/extensions/xep-0480.html

**Comment by naqvis** [comment]
> Can you look for XEP-0480: SASL Upgrade Tasks support?
> 
>     * https://xmpp.org/extensions/xep-0480.html

@Neustradamus Added `XEP-0388` and `XEP-0480` and made many other enhancements. Changelog,  release notes or protocol.md should provide much better change description

**Comment by Neustradamus** [comment]
@naqvis: Sorry for the delay!

A new XEP has been created, the XEP-0515, can you add it?

Except this one, Channel Binding is fully supported?

Linked to:
- RFC 5929: Channel Bindings for TLS: https://datatracker.ietf.org/doc/html/rfc5929
- RFC 9266: Channel Bindings for TLS 1.3: https://datatracker.ietf.org/doc/html/rfc9266
- XEP-0388: Extensible SASL Profile: https://xmpp.org/extensions/xep-0388.html
- XEP-0440: SASL Channel-Binding Type Capability: https://xmpp.org/extensions/xep-0440.html
- XEP-0474: SASL SCRAM Downgrade Protection: https://xmpp.org/extensions/xep-0474.html
- XEP-0480: SASL Upgrade Tasks: https://xmpp.org/extensions/xep-0480.html
- XEP-0515: TLS Channel-Binding Downgrade Protection: https://xmpp.org/extensions/xep-0515.html

**Comment by naqvis** [comment]
> A new XEP has been created, the XEP-0515, can you add it?

yeah, just pushed that XEP support.

> Except this one, Channel Binding is fully supported?

yes, all of them are implemented. refer to PROTOCOL.md for all supported extensions and their respective versions.
